ambry-store/src/main/java/com.github.ambry.store/CompactionLog.java

```diff
@@ -0,0 +1,336 @@
+   */
+  void setSafeToken(StoreFindToken safeToken) {
+    CycleLog cycleLog = getCurrentCycleLog();
+    if (!cycleLog.getState().equals(State.COPY)) {
+      throw new IllegalStateException("Cannot set a safe token - not in COPY phase");
+    }
+    cycleLog.safeToken = safeToken;
+    flush();
+  }
+
+  /**
+   * Marks the start of the copy phase.
+   */
+  void markCopyStart() {
+    getCurrentCycleLog().copyStartTime = time.milliseconds();
+    flush();
+  }
+
+  /**
+   * Marks the start of the switch phase.
+   */
+  void markSwitchStart() {
+    getCurrentCycleLog().switchStartTime = time.milliseconds();
+    flush();
+  }
+
+  /**
+   * Marks the start of the cleanup phase.
+   */
+  void markCleanupStart() {
+    getCurrentCycleLog().cleanupStartTime = time.milliseconds();
+    flush();
+  }
+
+  /**
+   * Marks the current compaction cycle as complete.
+   */
+  void markCycleComplete() {
+    getCurrentCycleLog().cycleEndTime = time.milliseconds();
+    currentIdx++;
+    flush();
+  }
+
+  /**
+   * Closes the compaction log. If compaction is complete, renames the log to keep a permanent record.
+   */
+  @Override
+  public void close() {
+    if (file.exists() && getCompactionState().equals(State.DONE)) {
+      File savedLog = new File(file.getAbsolutePath() + "_" + startTime);
+      if (!file.renameTo(savedLog)) {
+        throw new IllegalStateException("Compaction log could not be renamed after completion of compaction");
+      }
+    }
+  }
+
+  /**
+   * @return the {@link CycleLog} for the current compaction cycle.
+   */
+  private CycleLog getCurrentCycleLog() {
+    return cycleLogs.get(currentIdx);
+  }
+
+  /**
+   * Flushes all changes to the file backing this compaction log.
+   */
+  private void flush() {
+    /*
+        Description of serialized format
+
+        version
+        startTime
+        index of current cycle's log
+        size of cycle log list
+        cycleLog1 (see CycleLog#toBytes())
+        cycleLog2
+        ...
+        crc
+       */
+    File tempFile = new File(file.getAbsolutePath() + ".tmp");
+    try (FileOutputStream fileOutputStream = new FileOutputStream(tempFile)) {
+      CrcOutputStream crcOutputStream = new CrcOutputStream(fileOutputStream);
+      DataOutputStream stream = new DataOutputStream(crcOutputStream);
+      stream.writeShort(VERSION_0);
+      stream.writeLong(startTime);
+      stream.writeInt(currentIdx);
+      stream.writeInt(cycleLogs.size());
+      for (CycleLog cycleLog : cycleLogs) {
+        stream.write(cycleLog.toBytes());
+      }
+      stream.writeLong(crcOutputStream.getValue());
+      fileOutputStream.getChannel().force(true);
+    } catch (IOException e) {
+      throw new IllegalStateException(e);
+    }
+    if (!tempFile.renameTo(file)) {
+      throw new IllegalStateException("Newly written compaction log could not be saved");
+    }
+  }
+
+  /**
+   * Details and log for a single compaction cycle.
```

Comment:
I believe that this might not be the right place to explain what a single compaction cycle is. It should perhaps be explained in the `CompactionManager` or the `BlobCompactor` when they are written.